### PR TITLE
Make the memory printing in TestDirectory more robust

### PR DESCRIPTION
**Pull request by stevelinton**
On 32 bit systems the memory count can easily overflow. This at
least avoid attempting to print negative amounts, although it
can still fail to detect if it has overflowed and wrapped round,
giving a result wrong by some multiple of 4GB.

### CODE REVIEW

**Comment by fingolfin** [comment]
So... why don't we simply always track the memory usage in an `UInt8` resp. `uint64_t`? These ought to work even on 32bit systems. And we already provide `ObjInt_Int8` to convert an `Int8` into a GAP integer object; it was added for `NanosecondsSinceEpochInfo`.

**Comment by stevelinton** [comment]
@fingolfin Ah that makes sense. My habits still come from the days when 32 bit systems did not necessarily have a 64 bit integer type at all.

**Comment by fingolfin** [comment]
All we support seem to have one. And I am fairly hopeful that as we move forward, this will not change; and even if, then we'll have to deal with it anyway :-)

**Comment by stevelinton** [comment]
I see that making storage totals uniformly UInt8 across the system (which seems like the sensible way to do this) would change (in a completely boring way) quite a lot of profiling code. @ChrisJefferson would it cause any subtle problems for the line-by-line profiling?

**Comment by ChrisJefferson** [comment]
It should be fine, we might need some tiny macro to generate the correct printf, to print out an `Int8`. If you try making the changes, I'll sanity check them.

Will also make sure the profiling package can read 64-bit numbers even on a 32-bit GAP, but actually that would probably always make sense.

**Comment by fingolfin** [comment]
Such macros are part of the C99 standard, see e.g. http://en.cppreference.com/w/c/types/integer for an overview. Here, you may want to use `PRId64`.

**Comment by fingolfin** [comment]
So... is anybody working on extending the storage size? Do we still want to merge this PR?

**Comment by stevelinton** [comment]
I'm working on it. Turns out I need functions "dual" to ObjInt_Int8 to convert GAP integers back into Int8 types, but this is not hard.

**Comment by stevelinton** [comment]
Fixed I think. Minor side effect is that the profiling overhead in 32 bit may be increased, as memory totals have to be stored in GMP ints rather than small ints. On the other hand, the results should be accurate.

File: lib/string.gi [hunk 1]
```diff
@@ -1210,6 +1210,9 @@ end);
 
 InstallGlobalFunction(StringOfMemoryAmount, function(m)
     local  whole, frac, shift, s, units;
+    if not IsInt(m) or m < 0 then
+        Error("StringOfMemoryAmount: amount must be a non-negative integer number of bytes");
+    fi;
```
Comment (fingolfin) [inline lib/string.gi]: Can this still happen?

File: src/gap.c [hunk 2]
```diff
@@ -357,8 +357,7 @@ Obj Shell ( Obj context,
       /* This should be correct for small allocated totals at least, 
          even on 32 bits, but this functionality will never be very useful
          on 32 bits */
```
Comment (fingolfin) [inline src/gap.c]: Isn't that comment outdated now?

File: src/gap.c [hunk 3]
```diff
@@ -1986,7 +1985,7 @@ Obj FuncSHALLOW_SIZE (
 */
 
 Obj FuncTotalMemoryAllocated( Obj self ) {
-    return INTOBJ_INT(SizeAllBags);
+    return ObjInt_Int8(SizeAllBags);
```
Comment (fingolfin) [inline src/gap.c]: But `SizeAllBags` is an uint, so this may cause an implicit cast warning later on.

Do you intentionally use `Int8` instead of `UInt8` here? E.g. to catch a hypothetical overflow?

File: src/gmpints.c [hunk 4]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
 #endif
 }
 
+/**************************************************************************
+**
+** The following functions a GAP integer into an Int, UInt, Int8 or UInt8
```
Comment (fingolfin) [inline src/gmpints.c]: Insert "convert" ?

File: src/gmpints.h [hunk 5]
```diff
@@ -70,6 +70,17 @@ extern Obj ObjInt_Int(Int i);
 extern Obj ObjInt_UInt(UInt i);
 extern Obj ObjInt_Int8(Int8 i);
 
+/**************************************************************************
+**
+** The following functions a GAP integer into an Int, UInt, Int8 or UInt8
```
Comment (fingolfin) [inline src/gmpints.h]: Do we really want to duplicate this comment (and then maintain it in two places)? Perhaps just add it here, in the header, and remove the copy in the .c file, resp. change the comment there to just say something to the effect of "For details, see the .h file"

File: src/gmpints.c [hunk 6]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+** if it is in range, even if it is too big to be a small integer
```
Comment (fingolfin) [inline src/gmpints.c]: Not sure why "small integer" is even mentioned here?

Also, state what happens if it is *not* in range?

File: src/gmpints.c [hunk 7]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+*/
+Int Int_ObjInt(Obj i) {
+    UInt sign = 0;
+    if (IS_INTOBJ(i))
+        return INT_INTOBJ(i);
+    // must be a single limb
+    assert( sizeof(mp_limb_t) == sizeof(Int));
```
Comment (fingolfin) [inline src/gmpints.c]: This should ideally be turned into a `GAP_STATIC_ASSERT`, and moved to the top of the file.

File: src/gmpints.c [hunk 8]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+        ErrorMayQuit("Conversion error, integer too large", 0L, 0L);
+    return sign ? -(Int)val : (Int)val;
+}
+
+UInt UInt_ObjInt(Obj i)
+{
+    if (IS_INTOBJ(i)) {
+        Int val = INT_INTOBJ(i);
+        if (val < 0)
+            ErrorMayQuit("Conversion: negative integer into unsigned type",0,0);
```
Comment (fingolfin) [inline src/gmpints.c]: Instead of having this error twice, you could just start the function like so:
```C
UInt UInt_ObjInt(Obj i)
{
  if (IS_NEGATIVE(i))
    ErrorMayQuit("Conversion: negative integer into unsigned type",0,0);
  if (IS_INTOBJ(i))
    return INT_INTOBJ(i);
  if (TNUM_BAG(i) != T_INTPOS)
    .. 
```

File: src/gmpints.c [hunk 9]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+        else
+            return (UInt)val;
+    }
+    if (TNUM_BAG(i) == T_INTNEG)
+        ErrorMayQuit("Conversion: negative integer into unsigned type",0,0);
+    if (TNUM_BAG(i) != T_INTPOS)
+        ErrorMayQuit("Conversion error, expecting an integer, not a %s",(Int)TNAM_OBJ(i),0);
+        
+    // must be a single limb
+    assert( sizeof(mp_limb_t) == sizeof(Int));
```
Comment (fingolfin) [inline src/gmpints.c]: That assert is redundant if you turn the first one into a `GAP_STATIC_ASSERT`

File: src/gmpints.c [hunk 10]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+    if (SIZE_BAG(i) != sizeof(mp_limb_t))
```
Comment (fingolfin) [inline src/gmpints.c]: Elsewhere in this file, we write `SIZE_INT(i) != 1` for similar tasks. Not saying you should use that here, but I'd like to keep this uniform (at least on the long run, i.e. if we merge the above line as-is, I'd later go over the file to change it and/or all similar places to something similar).

File: src/gmpints.c (hunk 8 above)
Comment (fingolfin) [reply src/gmpints.c]: Of course with PR #1903, `IS_NEGATIVE` is renamed to `IS_NEG_INT` -- it's a pity these two PRs are clashing now. 

Since I am probably less annoyed by adapting, I guess we can merge this PR first and then I'll adapt PR #1903. *shrug*

File: src/gmpints.c [hunk 11]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+Int8 Int8_ObjInt(Obj i) {
+#ifndef SYS_IS_64_BIT
+    // in this case Int8 is Int
+    return Int_ObjInt(i);
+#else
+    UInt sign;
+    if (IS_INTOBJ(i))
+        return (Int8) INT_INTOBJ(i);
+    // must be at most two limbs
+    assert( 2*sizeof(mp_limb_t) == sizeof(Int8));
```
Comment (fingolfin) [inline src/gmpints.c]: Another `GAP_STATIC_ASSERT` candidate...

But wait, that seems wrong: this is the 64bit branch, so I'd assume that assert to always fire. Is it being tested? Or am I just misunderstanding something?

File: src/gmpints.c [hunk 12]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+    // must be a single limb
+    assert( sizeof(mp_limb_t) == sizeof(Int));
+    if (SIZE_BAG(i) != sizeof(mp_limb_t))
+        ErrorMayQuit("Conversion error, integer too large", 0L, 0L);
+    return ADDR_INT(i)[0];
+}
+
+Int8 Int8_ObjInt(Obj i) {
+#ifndef SYS_IS_64_BIT
+    // in this case Int8 is Int
```
Comment (fingolfin) [inline src/gmpints.c]: Err... no.. perhaps you mean `#ifdef` instead of `#ifndef`?

File: src/gmpints.c [hunk 13]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+        (sign && (val > (UInt8)INT64_MIN))) // works because of twos complement arithmetic
+        ErrorMayQuit("Conversion error, integer too large", 0L, 0L);
+    return sign ? -(Int8)val : (Int8)val;
+#endif
+}
+
+UInt8 UInt8_ObjInt(Obj i)
+{
+#ifndef SYS_IS_64_BIT
+    // in this case UInt8 is UInt
```
Comment (fingolfin) [inline src/gmpints.c]: See above...

File: src/gmpints.c [hunk 14]
```diff
@@ -568,6 +568,114 @@ Obj ObjInt_Int8( Int8 i )
...
+    return UInt_ObjInt(i);
+#else
+    if (IS_INTOBJ(i)) {
+            Int val = INT_INTOBJ(i);
```
Comment (fingolfin) [inline src/gmpints.c]: Weird indentation (but also: could be adjusted to use `IS_NEGATIVE`, as described above)

**Comment by fingolfin** [comment]
The tests pass, yet I think the code contains a 32 vs 64 bit mixup. If I am correct on that, this means the code is not covered suitably by our tests, which should ideally also be addressed (I know that we had several bugs in this conversion functions before, so I'd really like to see some kind of test, ideally covering "corner cases")

**Comment by fingolfin** [comment]
See also https://codecov.io/gh/gap-system/gap/pull/1909/src/src/gmpints.c?before=src/gmpints.c for more on the coverage

File: lib/string.gi (hunk 1 above)
Comment (stevelinton) [reply lib/string.gi]: This is a GAP callable function. Someone could pass something stupid

File: src/gap.c (hunk 3 above)
Comment (stevelinton) [reply src/gap.c]: There is no ObjInt_UInt8 for some reason. I guess I'll add it.

File: lib/string.gi (hunk 1 above)
Comment (fingolfin) [reply lib/string.gi]: Right, sorry, I meant to put this in another spot :)

File: lib/test.gi [hunk 15]
```diff
@@ -584,7 +584,16 @@ InstallGlobalFunction( "TestDirectory", function(arg)
       g := GASMAN_STATS();    
       return g[1][8] + g[2][8];    
   end;
-        
+  
+  SafeStringOfMemoryAmount := function(m)
+      if m < 0 then 
+          return "<overflow>";
```
Comment (fingolfin) [inline lib/test.gi]: Can this still happen?

**Comment by stevelinton** [comment]
I'm picking up the small things. The issue with tests is that the C functions are not used anywhere. I added them because it seemed logical to do all four cases if I was doing one, but in fact two of them are never called. It seems strange to add an artificial way to call them, so perhaps I should just add comments to the ones that are never called saying that they are not currently used or tested.

**Comment by fingolfin** [comment]
I don't like code that is not called by anything. It tends to bitrot, and there is always a risk that you get it wrong, too.

So I really am reluctant to add it in at all, except perhaps commented out. Or with tests that at least proof they work OK in some cases. Ideally, we'd have a setup to add test for functions on the C level (I think that would be very useful, also beyond this PR), but of course it is unreasonable to expect such a thing as part of this PR.
So the next best that comes to mind is indeed to add an "artificial" test function, callable from GAP. That is somewhat "artificial", sure, but then, aren't most if not all our tests "artificial" to some degree, too?

Actually, I am kind of dismayed to see we don't have tests for `ObjInt_Int`, `ObjInt_Int8` and `ObjInt_UInt` -- I thought I had added some when I worked on them. But looking at 547137cac9950326afdecf5ab393f23ab980a6b8 I see that I talk about a test setup, but I did not commit it anywhere :-/. I regret that.

Anyway, since I also failed to add test for these functions back when I touched them, I will not be so hypocritical as to block this PR over such tests :-). But I'll file an issue about this as a reminder :-)

File: lib/test.gi [hunk 16]
```diff
@@ -572,7 +572,7 @@ InstallGlobalFunction( "TestDirectory", function(arg)
            showProgress, suppressStatusMessage, exitGAP, c, files, 
            filetimes, filemems, recurseFiles, f, i, startTime, 
            startMem, testResult, time, mem, startGcTime, gctime,
-           totalGcTime, filegctimes, GcTime;
+           totalGcTime, filegctimes, GcTime, SafeStringOfMemoryAmount;
```
Comment (fingolfin) [inline lib/test.gi]: `SafeStringOfMemoryAmount` can now also go.

File: lib/test.gi [hunk 17]
```diff
@@ -584,7 +584,7 @@ InstallGlobalFunction( "TestDirectory", function(arg)
       g := GASMAN_STATS();    
       return g[1][8] + g[2][8];    
   end;
-        
+
```
Comment (fingolfin) [inline lib/test.gi]: Also this line can go. And I see that your first commit adds stuff here, and then some other commit removes it again. The first commit (and its message) are now almost pointless. Perhaps squash everything together? Or perhaps you want the change in `lib/string.gi` to be in a separate commit?

You could do this:
```sh
git checkout THIS-BRANCH
git rebase master # to make sure you really sit on master
git reset HEAD # discard all commits, but keep their changes

# now group changes into fresh commits:
git commit -m "Validate argument to StringOfMemoryAmount" lib/string.gi

# rest can probably be in a single big commit?
```

File: src/gmpints.c [hunk 18]
```diff
@@ -568,6 +571,117 @@ Obj ObjInt_Int8( Int8 i )
...
+  }
+
+  /* we need two limbs to store this integer */
+  assert( sizeof(mp_limb_t) == 4 );
+  Obj gmp = NewBag( T_INTPOS, 2 * sizeof(mp_limb_t) );
+  mp_limb_t *ptr = ADDR_INT(gmp);
+  ptr[0] = (UInt4)i;
+  ptr[1] = ((UInt8)i) >> 32;
+  return gmp;
+#endif
```
Comment (fingolfin) [inline src/gmpints.c]: Maybe run `git clang-format` ?

Comment (fingolfin) [reply src/gmpints.c]: (Then again, the rest of this file isn't formatted that way either *shrug*)

File: src/gmpints.c [hunk 19]
```diff
@@ -568,6 +571,117 @@ Obj ObjInt_Int8( Int8 i )
...
+    if (SIZE_BAG(i) != sizeof(mp_limb_t))
+        ErrorMayQuit("Conversion error, integer too large", 0L, 0L);
+    UInt val = ADDR_INT(i)[0];
+#ifdef SYS_IS_64_BIT
+    if ((!sign && (val > INT64_MAX)) ||
+        (sign && (val > (UInt)INT64_MIN))) // works because of twos complement arithmetic
+#else
+    if ((!sign && (val > INT32_MAX)) ||
+        (sign && (val > (UInt)INT32_MIN))) // works because of twos complement arithmetic
+#endif
```
Comment (fingolfin) [inline src/gmpints.c]: This is, "of course" an overflow check. But not everybody may know that, and if you don't, then this code looks like arcane magic, and the "works because ..." comment is not helpful either, because the reader may not even know what is supposed to "work" here.

In short: perhaps prefix this with a comment like `// check whether the result will fit into an Int`?

File: src/gmpints.c [hunk 20]
```diff
@@ -568,6 +571,117 @@ Obj ObjInt_Int8( Int8 i )
...
+    else
+        ErrorMayQuit("Conversion error, expecting an integer, not a %s",(Int)TNAM_OBJ(i),0);
+    
+    if (SIZE_INT(i) > 2)
+        ErrorMayQuit("Conversion error, integer too large", 0L, 0L);
+    UInt vall = ADDR_INT(i)[0];
+    UInt valh = (SIZE_INT(i) == 1) ?  0 : ADDR_INT(i)[1];
+    UInt8 val = (UInt8)vall + ((UInt8)valh <<32);
+    if ((!sign && (val > INT64_MAX)) ||
+        (sign && (val > (UInt8)INT64_MIN))) // works because of twos complement arithmetic
```
Comment (fingolfin) [inline src/gmpints.c]: See above
